From 053b2a50c14a3cfae8db7ae8fec0c6e675da66dd Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 12 Jun 2021 09:09:42 +0200
Subject: [PATCH 0733/1010] evl/irq: fix SMP kernel boot on uniprocessor
 machine

We may be running a SMP kernel on a uniprocessor machine whose
interrupt controller supports no IPI. We should attempt to hook IPIs
only if the hardware can support multiple CPUs, otherwise it is
unneeded and poised to fail.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/sched/core.c |  9 +++++++--
 kernel/evl/tick.c       | 33 ++++++++++++++++++---------------
 2 files changed, 25 insertions(+), 17 deletions(-)

diff --git a/kernel/evl/sched/core.c b/kernel/evl/sched/core.c
index dcda37f09b..5b4720e0de 100644
--- a/kernel/evl/sched/core.c
+++ b/kernel/evl/sched/core.c
@@ -719,6 +719,7 @@ static inline void leave_inband(struct evl_thread *root)
 #endif
 }
 
+#ifdef CONFIG_SMP
 /* oob stalled. */
 static irqreturn_t oob_reschedule_interrupt(int irq, void *dev_id)
 {
@@ -728,6 +729,9 @@ static irqreturn_t oob_reschedule_interrupt(int irq, void *dev_id)
 
 	return IRQ_HANDLED;
 }
+#else
+#define oob_reschedule_interrupt  NULL
+#endif
 
 static inline void set_next_running(struct evl_rq *rq,
 				struct evl_thread *next)
@@ -1347,7 +1351,8 @@ int __init evl_init_sched(void)
 		init_rq(rq, cpu);
 	}
 
-	if (IS_ENABLED(CONFIG_SMP)) {
+	/* See comment about hooking TIMER_OOB_IPI. */
+	if (IS_ENABLED(CONFIG_SMP) && num_possible_cpus() > 1) {
 		ret = __request_percpu_irq(RESCHEDULE_OOB_IPI,
 					oob_reschedule_interrupt,
 					IRQF_OOB,
@@ -1373,7 +1378,7 @@ void __init evl_cleanup_sched(void)
 	struct evl_rq *rq;
 	int cpu;
 
-	if (IS_ENABLED(CONFIG_SMP))
+	if (IS_ENABLED(CONFIG_SMP) && num_possible_cpus() > 1)
 		free_percpu_irq(RESCHEDULE_OOB_IPI, &evl_machine_cpudata);
 
 	for_each_online_cpu(cpu) {
diff --git a/kernel/evl/tick.c b/kernel/evl/tick.c
index c622b287f0..46ac430ef2 100644
--- a/kernel/evl/tick.c
+++ b/kernel/evl/tick.c
@@ -81,14 +81,14 @@ static int proxy_set_oneshot_stopped(struct clock_event_device *proxy_dev)
 }
 
 #ifdef CONFIG_SMP
-
 static irqreturn_t clock_ipi_handler(int irq, void *dev_id)
 {
 	evl_core_tick(NULL);
 
 	return IRQ_HANDLED;
 }
-
+#else
+#define clock_ipi_handler  NULL
 #endif
 
 static void setup_proxy(struct clock_proxy_device *dev)
@@ -108,14 +108,20 @@ int evl_enable_tick(void)
 {
 	int ret;
 
-#ifdef CONFIG_SMP
-	ret = __request_percpu_irq(TIMER_OOB_IPI,
-				clock_ipi_handler,
-				IRQF_OOB, "EVL timer IPI",
-				&evl_machine_cpudata);
-	if (ret)
-		return ret;
-#endif
+	/*
+	 * We may be running a SMP kernel on a uniprocessor machine
+	 * whose interrupt controller provides no IPI: attempt to hook
+	 * the timer IPI only if the hardware can support multiple
+	 * CPUs.
+	 */
+	if (IS_ENABLED(CONFIG_SMP) && num_possible_cpus() > 1) {
+		ret = __request_percpu_irq(TIMER_OOB_IPI,
+					clock_ipi_handler,
+					IRQF_OOB, "EVL timer IPI",
+					&evl_machine_cpudata);
+		if (ret)
+			return ret;
+	}
 
 	/*
 	 * CAUTION:
@@ -129,11 +135,8 @@ int evl_enable_tick(void)
 	 * device supports oneshot mode, or fails.
 	 */
 	ret = tick_install_proxy(setup_proxy, &evl_oob_cpus);
-	if (ret) {
-#ifdef CONFIG_SMP
-		free_percpu_irq(TIMER_OOB_IPI,
-				&evl_machine_cpudata);
-#endif
+	if (ret && IS_ENABLED(CONFIG_SMP) && num_possible_cpus() > 1) {
+		free_percpu_irq(TIMER_OOB_IPI, &evl_machine_cpudata);
 		return ret;
 	}
 
-- 
2.30.2

